homeassistant/components/stream/core.py

```diff
@@ -42,17 +63,166 @@ class Segment:
     def complete(self) -> bool:
         """Return whether the Segment is complete."""
         return self.duration > 0
 
-    def get_bytes_without_init(self) -> bytes:
+    @property
+    def data_size_with_init(self) -> int:
+        """Return the size of all part data + init in bytes."""
+        return len(self.init) + self.data_size
+
+    @property
+    def data_size(self) -> int:
+        """Return the size of all part data without init in bytes."""
+        # We can use the last part to quickly calculate the total data size.
+        if not self.parts_by_byterange:
+            return 0
+        last_http_range_start, last_part = next(
+            reversed(self.parts_by_byterange.items())
+        )
+        return last_http_range_start + len(last_part.data)
+
+    @callback
+    def async_add_part(
+        self,
+        part: Part,
+        duration: float,
+    ) -> None:
+        """Add a part to the Segment.
+
+        Duration is non zero only for the last part.
+        """
+        self.parts_by_byterange[self.data_size] = part
+        self.duration = duration
+        for output in self._stream_outputs:
+            output.part_put()
+
+    def get_data(self) -> bytes:
         """Return reconstructed data for all parts as bytes, without init."""
-        return b"".join([part.data for part in self.parts])
+        return b"".join([part.data for part in self.parts_by_byterange.values()])
+
+    def get_aggregating_bytes(
+        self, start_loc: int, end_loc: int | float
+    ) -> Generator[bytes, None, None]:
+        """Yield available remaining data until segment is complete or end_loc is reached.
+
+        Begin at start_loc. End at end_loc (exclusive).
+        Used to help serve a range request on a segment.
+        """
+        pos = start_loc
+        while (part := self.parts_by_byterange.get(pos)) or not self.complete:
+            if not part:
+                yield b""
+                continue
+            pos += len(part.data)
+            # Check stopping condition and trim output if necessary
+            if pos >= end_loc:
+                assert isinstance(end_loc, int)
+                # Trimming is probably not necessary, but it doesn't hurt
+                yield part.data[: len(part.data) + end_loc - pos]
+                return
+            yield part.data
+
+    def _render_hls_template(self, last_stream_id: int, render_parts: bool) -> str:
+        """Render the HLS playlist section for the Segment.
+
+        The Segment may still be in progress.
+        This method stores intermediate data in hls_playlist_parts, hls_num_parts_rendered,
+        and hls_playlist_complete to avoid redoing work on subsequent calls.
+        """
+        if self.hls_playlist_complete:
+            return self.hls_playlist_template
+        # Promote str to list[str] to allow appends and joins
+        playlist_template = (
+            [self.hls_playlist_template] if self.hls_playlist_template else []
```

Comment:
`hls_playlist_template` is actually a string, but we use a list during the function. If we initialize it to `[]` we can avoid the conversion on the first call (but might still have to test the condition) but will still need to do this conversion on each subsequent call.
Maybe a better alternative is to just store this as a `list[str]` and leave it to the consumers to join it into `str` when accessed.

Reply:
I just did this but realized I tried it before. The problem with the `list[str]` approach is that we still want to concatenate the strings every time the function is called, but the `"".join() `call will return an empty string, and you need to test for that before putting the joined result back into a new list. So instead of doing this test at the top of the function, you end up doing something similar at the bottom of the function when storing the intermediate data again.

Reply:
Actually what I mentioned above is not quite the issue. `hls_playlist_template` should not end up with an empty str after one call to the function. The issue is more with the parts, as it is possible to be end up with no parts after a call, which would result in a value of `[""]`. However, an additional "" in the list will result in an extra newline after joining which will mess up the pretty print. Note that this is not really an issue for the RFC as blank lines are disregarded, but let me see if I can work around that.

Reply:
Added a commit that addresses this